### ISSUE TITLE
Clusterctl topology dry run

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/cluster-api#5319]
**User Story**

As a user/operator I would like to have a tooling that helps me migrating my templates to clusterclass
As a user/operator I would like to have a tooling that helps me in designing new clusterclasses
As a user/operator I would like to have a tooling that helps me in understanding what are the effects of a change in a clusterclass

**Detailed Description**

ClusterClass and managed topology greatly simplify UX for end users.

However, as a ClusterClass author it would be great to have some tooling to test how a topology generated starting from ClusterClass looks like.

The first idea to explore in this space is to introduce a new clusterctl command, that takes in input a file with a Cluster, a ClusterClass and related templates and returns the resulting managed topology.

Once this initial prototype is ready, we can expand iterate on the UX, e.g. allowing to use already existing Cluster/ClusterClasses or templates so we can dry run not only the initial provisioning but also changes on existing Cluster.

/kind feature

**Comment by fabriziopandini** [comment]
/area topology
/milestone v1..0

/assign
For the initial exploration work, but then I will probably end over the work to @ykakarap if he wants to step in

**Comment by ykakarap** [comment]
@fabriziopandini I am happy to take it up! 😃

**Comment by fabriziopandini** [comment]
Quick update:
We are going to enhance the WIP PR in order to get new alpha commands that support a couple of use cases derived from the ones in the issue above:

- I would like to test my ClusterClass without actually creating objects
- I would like to test what happens if I change topology values on an existing Cluster
- I would like to test what happens if I do rebase on an existing Cluster (this is a special case of the previous point)
- I would like to test when happens if I change a ClusterClass
- I would like to test when happens if I change templates referenced by a ClusterClass

The following considerations apply to all the use cases above:
- Users is will be required to provide modified objects via a file; the command will try to read everything else from the Management Cluster if any.
- We are considering if/how to apply validation and defaulting rules to the provided CAPI objects; please note that defaulting/validation of provider-specific objects is not possible at this stage, and is thus out of scope
- The command will be limited to handle one modified Cluster and/or one modified ClusterClass max at a time in the changeset (the file).
- Changes to templates will be considered only if the template is referenced by a ClusterClass included in the changeset (the file).
- In case the change impact many Clusters, the list of impacted Clusters will be provided and the users will be required to pick one to get more detailed results about the operation effects.
- Detailed results (e.g object YAML or object diff) will be stored in a folder, while the command output will always be a summary of the changes introduced by the operation (e.g. Number of objects created, modified, deleted, etc.)
- The internal library method supporting the dry run operation will provide only raw objects as an output, thus allowing the consumers to decide how to surface relevant info to the users.

@ykakarap feel free to correct me if something is incomplete/incorrect
